### PR TITLE
Adjust tests to be able to run on any supplied RWX Filesystem storage

### PR DESCRIPTION
**Pull request by akalenyu**
**What this PR does / why we need it**:
Currently, we have tests that create an NFS server pod from code, pinning a bunch of tests to NFS instead of 
being able to run on any supplied RWX Filesystem storage.

Using infrastructure that is in place (`"storageRWXFileSystem": "nfs-csi"`) and the fact that we have NFS CSI in our toolbox, these tests can be generalized to work with any storage we throw at them (cephfs, for example).

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
- There are still usages of `InitNFS()` that remain which test using the NFS ipv6 address - should make a decision on those.
- Thanks @xpivarc for helping with migrations tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: tests/migration_test.go
Comment (akalenyu) [reply tests/migration_test.go:1565]: Done

File: tests/default-config.json
Comment (akalenyu) [reply tests/default-config.json]: Yep, no reason to deploy NFS CSI on all lanes
migration and storage use the ceph config

File: tests/storage/datavolume.go
Comment (akalenyu) [reply tests/storage/datavolume.go]: Done

File: tests/utils.go
Comment (akalenyu) [reply tests/utils.go]: I don't, removed

File: tests/utils.go
Comment (akalenyu) [reply tests/utils.go]: I don't, removed

File: tests/utils.go
Comment (akalenyu) [reply tests/utils.go:3130]: I would rather have this function receive the entire PVC object, it can easily require more info about the PVC

File: automation/test.sh
Comment (xpivarc) [inline automation/test.sh:61]: I now see that we are mixing `KUBEVIRT_STORAGE` into [focus settings](https://github.com/kubevirt/kubevirt/blob/main/automation/test.sh#L374). Would you mind moving it here? (we can of course do it in follow up)

File: tests/storage/datavolume.go
Comment (xpivarc) [reply tests/storage/datavolume.go:168]: Yes you need it but if you first wait for the DV to be Succeeded 
`				By("waiting for the dv import to pvc to finish")
				Eventually(ThisDV(dv), 180*time.Second).Should(HaveSucceeded())`
				
then you can do simple Get on the PVC, right?

**Review by xpivarc — APPROVED** [review]
Great improvement, I have some minor comments that can go as a follow-up.
/lgtm

I am going to try to convince you that you should drop commit `Address review comments` and integrate changes into appropriate commits(e.g `git commit --amend`). The commit will appear in the main as we are not squashing commits and doesn't necessarily bring value without looking up this PR and comments that were made.

File: automation/test.sh
Comment (akalenyu) [reply automation/test.sh:61]: Wouldn't this already be too late? (cluster-up already happened)

**Comment by akalenyu** [comment]
> Great improvement, I have some minor comments that can go as a follow-up. /lgtm
> 
> I am going to try to convince you that you should drop commit `Address review comments` and integrate changes into appropriate commits(e.g `git commit --amend`). The commit will appear in the main as we are not squashing commits and doesn't necessarily bring value without looking up this PR and comments that were made.

Took the offer :)

File: tests/storage/datavolume.go
Comment (akalenyu) [reply tests/storage/datavolume.go:168]: Yeah, right, but eventually is anyway a little more on the safe side so we don't flake out over a connection err

File: automation/test.sh
Comment (xpivarc) [reply automation/test.sh:61]: I was referring to this https://github.com/kubevirt/kubevirt/blob/main/automation/test.sh#L374. I think it's too late to have it there. Not sure how it is working. @dhiller any idea?

Comment (dhiller) [reply automation/test.sh:61]: @xpivarc you are absolutely right, L374 will not have any effect IMHO, as it's already after `make cluster-up`. We should move it here.

Comment (dhiller) [reply automation/test.sh:61]: Either in a followup PR or creating an issue would both be fine with me.

Comment (akalenyu) [reply automation/test.sh:61]: After looking at `KUBEVIRT_STORAGE` occurrences, I believe `export KUBEVIRT_STORAGE="rook-ceph"` can be safely removed here.
The migration job will have rook-ceph-default exported already and we match on wildcard in https://github.com/kubevirt/kubevirt/blob/ddc8e8ffbdada25b350580a403143e8932bdd9de/hack/functests.sh#L38-L40

File: automation/test.sh
Comment (xpivarc) [inline automation/test.sh:374]: @rmohr FYI

File: tests/migration_test.go
Comment (dhiller) [inline tests/migration_test.go:1525]: /hold

I don't see this test executed, both the 1.23 sig compute lane run and the 1.23 sig compute migrations show this test as skipped, same for the one below. I'd expect these tests to get executed in either lane.

I suspect that this test is skipped now due to missing test expectations. @akalenyu, can you please take a look why the tests are getting skipped?

File: automation/test.sh
Comment (akalenyu) [reply automation/test.sh]: Yeah @dhiller @xpivarc I was wrong, this has implications

Comment (xpivarc) [reply automation/test.sh]: I apologize for my blindness. Can you point to the exact place?

Comment (akalenyu) [reply automation/test.sh]: https://github.com/kubevirt/kubevirt/blob/5a734f92980248744ef02949a875cfe02c1c1682/automation/test.sh#L394-L399

File: tests/migration_test.go
Comment (akalenyu) [reply tests/migration_test.go:1525]: Thanks for catching this. discussing reasoning & ways forward [above](https://github.com/kubevirt/kubevirt/pull/7324#discussion_r830621517)

File: automation/test.sh
Comment (akalenyu) [reply automation/test.sh:374]: Now, instead of overriding the focus var, I am appending storage-req to it. WDYT? @dhiller @xpivarc

Comment (xpivarc) [reply automation/test.sh:374]: `if [[ ""rook-ceph"" == "rook-ceph-default" ]]; then` ?

**Review by rmohr — COMMENTED** [review]
I think that we used nfs is from the historic fact that we did not have rook-ceph enabled and stable on so many clusters. Right now we have ceph always deployed on the migrations lanes and the storage lanes. We could probably just ensure that the tests are moved to the right CI lanes and remove all nfs traces. Would that be worth a try?

**Comment by rmohr** [comment]
> > That is what I am trying to achieve here; switching the storage that is used for these (previously-NFS) tests should be as easy as changing this
> 
> Yes. I mostly meant: Why do we introduce this nfs-csi and not just use ceph for filesystem pvcs.

And maybe I missed it but I don't see the nfs code removed (like creating the server pod and so forth).

**Comment by akalenyu** [comment]
> > > I think that we used nfs is from the historic fact that we did not have rook-ceph enabled and stable on so many clusters. Right now we have ceph always deployed on the migrations lanes and the storage lanes. We could probably just ensure that the tests are moved to the right CI lanes and remove all nfs traces. Would that be worth a try?
> > 
> > 
> > That is what I am trying to achieve here; switching the storage that is used for these (previously-NFS) tests should be as easy as changing this
> 
> Yes. I mostly meant: Why do we introduce this nfs-csi and not just use ceph for filesystem pvcs.

That is because ceph rbd doesn't do RWX Filesystems PVCs (only Block).
Cephfs does, and if we ever pick up testing cephfs it should be as simple as overriding nfs-csi in the config

**Comment by akalenyu** [comment]
> > > That is because ceph rbd doesn't do RWX Filesystems PVCs (only Block).
> > > Cephfs does, and if we ever pick up testing cephfs it should be as simple as overriding nfs-csi in the config
> > 
> > 
> > Adding cephfs support would have been adding a storage class and a filesystem config yaml. I am not sure why this nfs-csi was introduced.
> 
> Let me just add that I think the direction of the PR is fine. I just don't think it makes much sense to introduce nfs-csi extra. It is still a remainder from past times where we did not have rook-ceph, then it did not work so well, and we had less lanes which had storage configured. We don't need nfs specifically.

I see. Is introducing cephfs that simple? I'd think we have to perform some kubevirtci prep similarly to NFS CSI.
Anyway, this PR enables shifting to another RWX FS storage by simply changing the config value.

We initially decided to introduce NFS CSI for the NFS CDI lane which was hitting some issues with manual creation/deletion of PVs, so thought it made sense to capitalize on having it here as well. 
But this PR is more about generalizing the tests to be able to work with any RWX FS storage.

BTW I kept the NFS functions because the IPv6 tests still use that.

**Comment by rmohr** [comment]
> BTW I kept the NFS functions because the IPv6 tests still use that.

I would love to remove our legacy nfs services with this PR too. Is there a specific reason why we can't switch these tests?

**Comment by akalenyu** [comment]
> > BTW I kept the NFS functions because the IPv6 tests still use that.
> 
> I would love to remove our legacy nfs services with this PR too. Is there a specific reason why we can't switch these tests?

Yes, testing an IPv6 address is tricky without this server pod

**Comment by xpivarc** [comment]
Can you drop 162c01df9bbbf1b1a5ad772afbaf3af38730fe4f in favor of https://github.com/kubevirt/kubevirt/pull/7450?

**Comment by akalenyu** [comment]
> Can you drop [162c01d](https://github.com/kubevirt/kubevirt/commit/162c01df9bbbf1b1a5ad772afbaf3af38730fe4f) in favor of #7450?

sure, done

**Comment by akalenyu** [comment]
> > > BTW I kept the NFS functions because the IPv6 tests still use that.
> > 
> > 
> > I would love to remove our legacy nfs services with this PR too. Is there a specific reason why we can't switch these tests?
> 
> Yes, testing an IPv6 address is tricky without this server pod plus they are pretty much NFS specific and cant be generalized, right?

@rmohr Is this blocked by https://github.com/kubevirt/kubevirtci/pull/768? 
We could always change the RWX FS storage that is to be used by tests and stop deploying NFS CSI in a follow up

**Comment by awels** [comment]
For @rmohr To better understand the reasoning behind some of the decisions here.
1. nfs-csi is a new storage provider introduced for CDI so we can have dynamic provisioning with NFS (no more statically creating PVs for tests). Since KubeVirt doesn't have cephfs enabled (yet) it makes sense to use the provider from kubevirtci, since its already there. It is NOT using the legacy NFS server at all. It is just providing RWX filesystem provisioner that is configured by the cluster setup.
2. We want to remove all NFS legacy code, as Alex pointed out there is some testing that seems to be testing that NFS supports IPv6, which IMO can simply be removed, but that would be out of the scope of this PR. We all agree it needs to be removed though.
3. The PR is about making RWX filesystem tests provisioner agnostic, so we can then pick whatever provisioner we want when we set up the cluster. We don't care if it is nfs-csi, cephfs, portworkx, or whatever else provides RWX filesystem volumes.

Once this is merged we can do a follow up for completely removing all the legacy NFS code.

**Comment by dhiller** [comment]
I see that the vgpu tests have passed once on the latest commit, this is another instance of the kind cluster not being able to delete, which we have encountered more often lately.

**Comment by dhiller** [comment]
Also I could verify that the test `test_id:2653` is actually run on the 1.23-sig-compute-migrations lane and passed.

**Comment by dhiller** [comment]
/approve
/override pull-kubevirt-e2e-kind-1.23-vgpu-nonroot

see [comment](https://github.com/kubevirt/kubevirt/pull/7324#issuecomment-1091613205)

**Comment by akalenyu** [comment]
/hold cancel

@dhiller thank you

**Pull request merged**
